source/mqtt_agent.c

```diff
@@ -718,6 +727,42 @@ static MQTTStatus_t createAndAddCommand( CommandType_t commandType,
 
 /*-----------------------------------------------------------*/
 
+static void concludeCommand( MQTTAgentContext_t * pAgentContext,
+                             Command_t * pCommand,
+                             MQTTStatus_t returnCode,
+                             uint8_t * pSubackCodes )
+{
+    bool commandReleased = false;
+    MQTTAgentReturnInfo_t returnInfo;
+
+    ( void ) memset( &returnInfo, 0x00, sizeof( MQTTAgentReturnInfo_t ) );
+    assert( pAgentContext != NULL );
+
+    returnInfo.returnCode = returnCode;
+    returnInfo.pSubackCodes = pSubackCodes;
+
+    if( pCommand != NULL )
+    {
+        CommandCallback_t pCallback = pCommand->pCommandCompleteCallback;
+
+        if( pCallback != NULL )
+        {
+            pCallback( pCommand->pCmdContext, &returnInfo );
+        }
+
+        commandReleased = pAgentContext->agentInterface.releaseCommand( pCommand );
+
+        if( !commandReleased )
+        {
+            LogError( ( "Command %p of type %d not released.",
```

Comment:
(Nit) I find it better to have the failure outcome be the subject of the statement
```suggestion
            LogError( ( "Failed to release command %p of type %d.",
```